Add Iommu related changes develop

prbasyal-amd opened the pull request:
1. Separated the Bar Overview section from ROCm and PCIe page and added into a new page.
2. Added IOMMU related text. 
3. Updated the style, formatting, and fixed grammatical errors without changing the legacy content.

prbasyal-amd committed:
Meta description updated

prbasyal-amd committed:
New page for Bar Memory created

prbasyal-amd committed:
Restructured the legacy table

prbasyal-amd committed:
New page for bar memory created, added IOMMU related changes, and implemented style guide in existing content

prbasyal-amd committed:
Prefetching added to word list

prbasyal-amd committed:
Prefetching added to word list

spolifroni-amd commented on docs/conceptual/Bar-Memory.rst:
```diff
@@ -0,0 +1,98 @@
+******************************
+BAR memory overview
+******************************
+From the BIOS of an Intel Xeon E5-based system, you can enable above 4 GB PCIe addressing. You also need to set the
```
The intro sounds like it's mid-conversation. It also doesn't mention BAR memory at all. I would change it to something that includes BAR memory and introduces the topic. I would pull the info from later up and put it up here:

"P2P DMA only works when one device can directly access the local BAR memory of another. If the BAR memory is above the physical addressing limit of the device, it will not be able to access the remote BAR."

and then say what ways are used and move on from there.

spolifroni-amd commented on docs/conceptual/Bar-Memory.rst:
```diff
@@ -0,0 +1,98 @@
+From the BIOS of an Intel Xeon E5-based system, you can enable above 4 GB PCIe addressing. You also need to set the
+memory-mapped input/output (MMIO) High Base address ``MMIOH Base`` and range ``MMIO High Size`` in the BIOS.
+
+In the Supermicro system, under the system BIOS, go to ``Advanced > PCIe/PCI/PnP configuration``. You should see the following:
```
It's unclear what these steps are meant to do/what problem they're meant to solve. They need to be contextualized.

spolifroni-amd commented on docs/conceptual/Bar-Memory.rst:
```diff
@@ -0,0 +1,98 @@
+
+.. Note:: 
+
+  When we support the Large Base Address Register (BAR) Capability there is a Large BAR VBIOS that disables the IO BAR.
```
The "when we support" feels weird. Either we support it or we don't. Or if we only support it in certain cases, we should say which. And we should also explain what the implications are of disabling the I/O BAR

prbasyal-amd committed:
New sections and review feedback from Sandra added

prbasyal-amd committed:
Profiler added to resolve merge conflict

prbasyal-amd committed:
Profiler added to resolve merge conflict

prbasyal-amd committed:
Merge branch 'develop' into add-IOMMU-related-changes-develop

spolifroni-amd commented on docs/conceptual/Bar-Memory.rst:
```diff
@@ -0,0 +1,97 @@
+* Enable the Input-Output Memory Management Unit (IOMMU). When the IOMMU is enabled in non-passthrough mode, it will create a virtual IO address space for each device on the system. It also ensures that all virtual addresses created in that space are within the physical addressing limits of the device. The driver reports the physical addressing limits to the kernel. The kernel sets the IO virtual ...
+
+
+BAR configuration for AMD GFX9 and Vega10 GPUs
```
This section is likely to go stale (and the Vega 10 is already pretty old), so maybe it should be removed from here and put somewhere with more visibility, like with installation.

peterjunpark committed:
Add minor stylistic updates in release notes (#4097)

danielsu-amd committed:
External CI: give manifest files same naming scheme as build artifacts (#4100)

randyh62 committed:
Update license.md (#4101)

Update ROCgdb license to point to GNU version 3.0

samjwu committed:
Update from ROCm 6.2.4 to 6.3.0 (#4103)

peterjunpark committed:
fix links to smi tools full changelog on GH (#4108)

dependabot[bot] committed:
Build(deps): Bump rocm-docs-core from 1.9.2 to 1.11.0 in /docs/sphinx (#4111)

Bumps [rocm-docs-core](https://github.com/ROCm/rocm-docs-core) from 1.9.2 to 1.11.0.
- [Release notes](https://github.com/ROCm/rocm-docs-core/releases)
- [Changelog](https://github.com/ROCm/rocm-docs-core/blob/develop/CHANGELOG.md)
- [Commits](https://github.com/ROCm/rocm-docs-core/compare/v1.9.2...v1.11.0)

---
updated-dependencies:
- dependency-name: rocm-docs-core
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

peterjunpark committed:
fix stack image (#4112)

danielsu-amd committed:
External CI: change rocm-core staging branch to master (#4115)

danielsu-amd committed:
External CI: create pipeline files for rocJPEG (#4117)

peterjunpark committed:
remove programming guide from TOC (#4116)

SwRaw committed:
Update what-is-rocm.rst (#4122)

prbasyal-amd committed:
Review feedback incorporated and moved the section to How-to

prbasyal-amd committed:
Reference in TOC and Index page updated

prbasyal-amd committed:
Changed the sentence flow

spolifroni-amd commented on docs/how-to/Bar-Memory.rst:
```diff
@@ -0,0 +1,102 @@
+
+Handling physical address limitation
+=============================================
+When the system boots, the system BIOS allocates the physical address space for different components in the system. This also includes system memory and one or more MMIO apertures. In general, there are two MMIO apertures: one allocated with below 4 GB of physical address space for 32-bit compatibility, and one with above 4 GB for devices that need more space. 
```
```suggestion
When a system boots, the system BIOS allocates the physical address space for the components in the system, including system memory and MMIO apertures. In general, there are two MMIO apertures: one allocated below 4 GB of physical address space for 32-bit compatibility, and one above 4 GB for devices that need more space. 
```

Is there a space between the 4 and GB? I feel like there isn't, but maybe there is.

spolifroni-amd commented on docs/how-to/Bar-Memory.rst:
```diff
@@ -0,0 +1,102 @@
+=============================================
+When the system boots, the system BIOS allocates the physical address space for different components in the system. This also includes system memory and one or more MMIO apertures. In general, there are two MMIO apertures: one allocated with below 4 GB of physical address space for 32-bit compatibility, and one with above 4 GB for devices that need more space. 
+
+You can control the memory address of the high MMIO aperture from the configuration options in the system BIOS. This enables you to configure the additional MMIO space to align with the physical addressing limit and allow P2P DMA between the devices. For example, if a PCIe device is limited to 44-bit of physical addressing, you should ensure that the MMIO aperture is set below 44-bit in the syste ...
```
```suggestion
You can control the memory address of the high MMIO aperture from the configuration options in the system BIOS. This lets you configure the additional MMIO space to align with the physical addressing limit and allows P2P DMA between the devices. For example, if a PCIe device is limited to 44-bit of physical addressing, you should ensure that the MMIO aperture is set below 44-bit in the system physical address space.
```

Maybe it's because this isn't my area of expertise, but I feel this may need a bit more explanation: 

> This lets you configure the additional MMIO space to align with the physical addressing limit and allow P2P DMA between the devices.

spolifroni-amd commented on docs/how-to/Bar-Memory.rst:
```diff
@@ -0,0 +1,102 @@
+
+There are two ways to handle this:
+
+* Ensure that the high MMIO aperture is within the physical addressing limits of the devices in the system. For example, if the devices have a 44-bit physical addressing limit, set the ``MMIOH Base`` and ``MMIO High size`` options in the BIOS such that the aperture is within the 44-bit address range. Also, ensure that the ``Above 4G Decoding`` option is Enabled. 
```
```suggestion
* Ensure that the high MMIO aperture is within the physical addressing limits of the devices in the system. For example, if the devices have a 44-bit physical addressing limit, set the ``MMIOH Base`` and ``MMIO High size`` options in the BIOS such that the aperture is within the 44-bit address range, and ensure that the ``Above 4G Decoding`` option is Enabled. 
```

spolifroni-amd commented on docs/how-to/Bar-Memory.rst:
```diff
@@ -0,0 +1,102 @@
+
+* Ensure that the high MMIO aperture is within the physical addressing limits of the devices in the system. For example, if the devices have a 44-bit physical addressing limit, set the ``MMIOH Base`` and ``MMIO High size`` options in the BIOS such that the aperture is within the 44-bit address range. Also, ensure that the ``Above 4G Decoding`` option is Enabled. 
+
+* Enable the Input-Output Memory Management Unit (IOMMU). When the IOMMU is enabled in the non-passthrough mode, it will create a virtual IO address space for each device on the system. It also ensures that all virtual addresses created in that space are within the physical addressing limits of the device. For more information on IOMMU, see :doc:`../conceptual/iommu`. 
```
```suggestion
* Enable the Input-Output Memory Management Unit (IOMMU). When the IOMMU is enabled in non-passthrough mode, it will create a virtual I/O address space for each device on the system. It also ensures that all virtual addresses created in that space are within the physical addressing limits of the device. For more information on IOMMU, see :doc:`../conceptual/iommu`. 
```

spolifroni-amd commented on docs/how-to/Bar-Memory.rst:
```diff
@@ -0,0 +1,102 @@
+    - Description
+  * - BAR0-1 registers
+    - 64-bit, Prefetchable, GPU memory
+    - 8 GB or 16 GB depending on GPU SKU. Set to less than 2^44 to support P2P access from other GPUs like Vega10. Prefetching enables faster read operation for high-performance computing (HPC) by fetching the contiguous data from the same data source even before requested as an anticipation of a future request.
```
I'm still concerned that this mentions the Vega and not the Instinct.

spolifroni-amd reviewed (changes requested):
Some places need clarification, but it's mostly just smoothing language at this point.

prbasyal-amd committed:
Feedback incorporated

lpaoletti commented on docs/how-to/Bar-Memory.rst:
```diff
@@ -0,0 +1,102 @@
+**************************************
+Troubleshoot BAR access limitation
+**************************************
+Direct Memory Access (DMA) to PCIe devices using Base Address Registers (BARs) can be restricted due to physical addressing limits. These restrictions can result in data access failures between the system components. Peer-to-peer (P2P) DMA is used to access resources such as registers and memory between devices. PCIe devices need memory-mapped input/output (MMIO) space for DMA and these MMIO spac ...
```
```suggestion
Direct Memory Access (DMA) to PCIe devices using Base Address Registers (BARs) can be restricted due to physical addressing limits. These restrictions can result in data access failures between the system components. Peer-to-peer (P2P) DMA is used to access resources such as registers and memory between devices. PCIe devices need memory-mapped input/output (MMIO) space for DMA, and these MMIO spaces are defined in the PCIe BARs. 
```

lpaoletti commented on docs/how-to/Bar-Memory.rst:
```diff
@@ -0,0 +1,102 @@
+**************************************
+Direct Memory Access (DMA) to PCIe devices using Base Address Registers (BARs) can be restricted due to physical addressing limits. These restrictions can result in data access failures between the system components. Peer-to-peer (P2P) DMA is used to access resources such as registers and memory between devices. PCIe devices need memory-mapped input/output (MMIO) space for DMA and these MMIO spac ...
+
+These BARs are a set of 32-bit or 64-bit registers that are used to define the memory space that a PCIe devices can access. These are also used by the CPU and other system devices to access the resources of the PCIe devices. P2P DMA only works when one device can directly access the local BAR memory of another. If the BAR memory is above the physical addressing limit of the device, the device wil ...
```
```suggestion
These BARs are a set of 32-bit or 64-bit registers that are used to define the memory space that PCIe devices can access. The CPU and other system devices also use these to access the resources of the PCIe devices. P2P DMA only works when one device can directly access the local BAR memory of another. If the BAR memory is above the physical addressing limit of the device, the device will not be able to access the remote BAR. 
```

lpaoletti commented on docs/how-to/Bar-Memory.rst:
```diff
@@ -0,0 +1,102 @@
+
+These BARs are a set of 32-bit or 64-bit registers that are used to define the memory space that a PCIe devices can access. These are also used by the CPU and other system devices to access the resources of the PCIe devices. P2P DMA only works when one device can directly access the local BAR memory of another. If the BAR memory is above the physical addressing limit of the device, the device wil ...
+
+To handle any BAR access issues that might occur, you need to be aware of the physical address limitations of the devices and understand the :ref:`BAR configuration of AMD GPUs <bar-configuration>`. This information is important when you need to set up additional MMIO apertures for PCIe devices in the physical address space of the system.
```
```suggestion
To handle any BAR access issues that might occur, you need to be aware of the physical address limitations of the devices and understand the :ref:`BAR configuration of AMD GPUs <bar-configuration>`. This information is important when you need to set up additional MMIO apertures for PCIe devices in the system's physical address space.
```

lpaoletti commented on docs/how-to/Bar-Memory.rst:
```diff
@@ -0,0 +1,102 @@
+
+Handling physical address limitation
+=============================================
+When a system boots, the system BIOS allocates the physical address space for the components in the system, including system memory and MMIO apertures. In general, there are two MMIO apertures: one allocated below 4 GB of physical address space for 32-bit compatibility, and one above 4 GB for devices that need more space.
```
```suggestion
When a system boots, the system BIOS allocates the physical address space for the components in the system, including system memory and MMIO apertures. There are two MMIO apertures: one allocated less than 4 GB of physical address space for 32-bit compatibility, and one more than 4 GB for devices needing more space.
```

lpaoletti commented on docs/how-to/Bar-Memory.rst:
```diff
@@ -0,0 +1,102 @@
+=============================================
+When a system boots, the system BIOS allocates the physical address space for the components in the system, including system memory and MMIO apertures. In general, there are two MMIO apertures: one allocated below 4 GB of physical address space for 32-bit compatibility, and one above 4 GB for devices that need more space.
+
+You can control the memory address of the high MMIO aperture from the configuration options in the system BIOS. This lets you configure the additional MMIO space to align with the physical addressing limit and allows P2P DMA between the devices. For example, if a PCIe device is limited to 44-bit of physical addressing, you should ensure that the MMIO aperture is set below 44-bit in the system phy ...
```
```suggestion
You can control the memory address of the high MMIO aperture from the system BIOS configuration options. This lets you configure the additional MMIO space to align with the physical addressing limit and allows P2P DMA between the devices. For example, if a PCIe device is limited to 44-bit of physical addressing, you should ensure that the MMIO aperture is set below 44-bit in the system physical address space.
```

lpaoletti commented on docs/how-to/Bar-Memory.rst:
```diff
@@ -0,0 +1,102 @@
+
+* Ensure that the high MMIO aperture is within the physical addressing limits of the devices in the system. For example, if the devices have a 44-bit physical addressing limit, set the ``MMIO High Base`` and ``MMIO High size`` options in the BIOS such that the aperture is within the 44-bit address range, and ensure that the ``Above 4G Decoding`` option is Enabled.  
+
+* Enable the Input-Output Memory Management Unit (IOMMU). When the IOMMU is enabled in the non-passthrough mode, it will create a virtual I/O address space for each device on the system. It also ensures that all virtual addresses created in that space are within the physical addressing limits of the device. For more information on IOMMU, see :doc:`../conceptual/iommu`. 
```
```suggestion
* Enable the Input-Output Memory Management Unit (IOMMU). When the IOMMU is enabled in non-passthrough mode, it will create a virtual I/O address space for each device on the system. It also ensures that all virtual addresses created in that space are within the physical addressing limits of the device. For more information on IOMMU, see :doc:`../conceptual/iommu`. 
```

lpaoletti commented on docs/how-to/Bar-Memory.rst:
```diff
@@ -0,0 +1,102 @@
+BAR configuration for AMD GPUs
+================================================
+
+For AMD GPUs that have the 44-bit physical address and 48-bit GPU virtual address limit, the BARs are configured as:
```
```suggestion
The following table shows how the BARs are configured for AMD GPUs that have a 44-bit physical address and 48-bit GPU virtual address limit. 
```

lpaoletti commented on docs/how-to/Bar-Memory.rst:
```diff
@@ -0,0 +1,102 @@
+    - Description
+  * - BAR0-1 registers
+    - 64-bit, Prefetchable, GPU memory
+    - 8 GB or 16 GB depending on GPU SKU. Set to less than 2^44 to support P2P access from other GPUs with 44-bit physical address limit. Prefetching enables faster read operation for high-performance computing (HPC) by fetching the contiguous data from the same data source even before requested as an anticipation of a future request.
```
```suggestion
    - 8 GB or 16 GB depending on GPU. Set to less than 2^44 to support P2P access from other GPUs with a 44-bit physical address limit. Prefetching enables faster read operation for high-performance computing (HPC) by fetching the contiguous data from the same data source even before requested as an anticipation of a future request.
```

lpaoletti commented on docs/how-to/Bar-Memory.rst:
```diff
@@ -0,0 +1,102 @@
+    - 8 GB or 16 GB depending on GPU SKU. Set to less than 2^44 to support P2P access from other GPUs with 44-bit physical address limit. Prefetching enables faster read operation for high-performance computing (HPC) by fetching the contiguous data from the same data source even before requested as an anticipation of a future request.
+  * - BAR2-3 registers
+    - 64-bit, Prefetchable, Doorbell
+    - Set to less than 2^44 to support P2P access from other GPUs with 44-bit physical address limit. As a Doorbell BAR, it indicates to the GPU that a new operation is in its queue to be processed. 
```
```suggestion
    - Set to less than 2^44 to support P2P access from other GPUs with a 44-bit physical address limit. As a Doorbell BAR, it indicates to the GPU that a new operation is in its queue to be processed. 
```

lpaoletti commented on docs/how-to/Bar-Memory.rst:
```diff
@@ -0,0 +1,102 @@
+
+1. GPU Frame Buffer BAR: ``Memory at bf40000000 (64-bit, prefetchable) [size=256M]``
+
+The size of the BAR in the example is 256 MB. In general, it will be the size of the
```
```suggestion
The size of the BAR in the example is 256 MB. Generally, it will be the size of the
```

lpaoletti commented on docs/how-to/Bar-Memory.rst:
```diff
@@ -0,0 +1,102 @@
+
+3. I/O BAR: ``I/O ports at 3000 [size=256]``
+
+This is for legacy VGA and boot device support. Since the GPUs used are not connected to a display (VGA devices), this is not a concern even if it is not set up in the SBIOS.
```
```suggestion
This is for legacy VGA and boot device support. Because the GPUs used are not connected to a display (VGA devices), this is not a concern, even if it isn't set up in the SBIOS.
```

lpaoletti replied:
SBIOS - this is correct? not BIOS?

prbasyal-amd replied:
system BIOS

lpaoletti commented on docs/how-to/Bar-Memory.rst:
```diff
@@ -0,0 +1,102 @@
+
+4. MMIO BAR: ``Memory at c7400000 (32-bit, non-prefetchable) [size=256K]``
+
+This is required by the AMD Driver SW to access the configuration registers. Since the reminder of the BAR available is only 1 DWORD (32-bit), this is set below 4 GB. In the example, it is fixed at 256 KB.
```
```suggestion
The AMD Driver SW requires this to access the configuration registers. Since the reminder of the BAR available is only 1 DWORD (32-bit), this is set less than 4 GB. In the example, it is fixed at 256 KB.
```

lpaoletti commented on docs/how-to/Bar-Memory.rst:
```diff
@@ -0,0 +1,102 @@
+
+5. Expansion ROM: ``Expansion ROM at c7440000 [disabled] [size=128K]``
+
+This is required by the AMD Driver SW to access the GPU video-bios. In the example, it is fixed at 128 KB.
```
SW - assume it means software - probably can remove
shouldn't **bios** be uppercase?

lpaoletti commented on docs/how-to/Bar-Memory.rst:
```diff
@@ -0,0 +1,102 @@
+5. Expansion ROM: ``Expansion ROM at c7440000 [disabled] [size=128K]``
+
+This is required by the AMD Driver SW to access the GPU video-bios. In the example, it is fixed at 128 KB.
+
```
remove spaces at the bottom

prbasyal-amd commented on docs/how-to/Bar-Memory.rst:
```diff
@@ -0,0 +1,102 @@
+**************************************
+Troubleshoot BAR access limitation
+**************************************
+A system can consist of multiple PCIe devices and data access between the devices is needed for proper functioning and optimum system performance. Direct Memory Access (DMA) to the device resources using the BARs (Base Address Registers) can be restricted due to the physical address limit. This restriction can result in data access failure between the system components. Peer-to-peer (P2P) DMA is  ...
```
Confirmed and updated

prbasyal-amd committed:
Leo's feedback incorporated

prbasyal-amd commented:
Due to merge conflicts and other file changes being tracked, all the review suggestions have been incorporated. Created a new PR for the final merge: https://github.com/ROCm/ROCm/pull/4168
